dubbo-common/src/main/java/org/apache/dubbo/common/URLBuilder.java

```diff
@@ -0,0 +1,350 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.common;
+
+import org.apache.dubbo.common.utils.CollectionUtils;
+import org.apache.dubbo.common.utils.StringUtils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public final class URLBuilder {
+    private String protocol;
+
+    private String username;
+
+    private String password;
+
+    // by default, host to registry
+    private String host;
+
+    // by default, port to registry
+    private int port;
+
+    private String path;
+
+    private Map<String, String> parameters;
+
+    public URLBuilder() {
+        protocol = null;
+        username = null;
+        password = null;
+        host = null;
+        port = 0;
+        path = null;
+        parameters = new HashMap<>();
+    }
+
+    public URLBuilder(String protocol, String host, int port) {
+        this(protocol, null, null, host, port, null, null);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String[] pairs) { // varargs ... conflict with the following path argument, use array instead.
+        this(protocol, null, null, host, port, null, CollectionUtils.toStringMap(pairs));
+    }
+
+    public URLBuilder(String protocol, String host, int port, Map<String, String> parameters) {
+        this(protocol, null, null, host, port, null, parameters);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path) {
+        this(protocol, null, null, host, port, path, null);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path, String... pairs) {
+        this(protocol, null, null, host, port, path, CollectionUtils.toStringMap(pairs));
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path, Map<String, String> parameters) {

```

Comment:
since it's builder, we should make all constructor **private**

Reply:
> since it's builder, we should make all constructor **private**

I don't think the constructor of the builder itself should be `private`, it is the constructor of the class that the builder is to build(the `URL` class) should be `private` to prevent client code from creating the (`URL`) objects without using the builder. But for compatibility, the constructor of `URL` should not be private at this moment.

About the visibilities of the constructors, please refer to [this thread](https://stackoverflow.com/a/37028913)

> This looks a bit strange because usually a builder uses private constructor to totally avoid the creation of the instance by the client code without using the builder.

and [this post](https://dzone.com/articles/design-patterns-the-builder-pattern):

> The Pattern
The Builder pattern allows us to write readable, understandable code to set up complex objects. It is often implemented with a fluent interface, which you may have seen in tools like Apache Camel or Hamcrest. The builder will contain all of the fields that exist on the BankAccount class itself. We will configure all of the fields that we want on the builder, and then we'll use the builder to create accounts. At the same time, we'll remove the public constructor from the BankAccount class and replace it with a private constructor so that accounts can only be created via the builder.

Reply:
@beiwei30 what do you think